crates/stages/src/stages/utils.rs

```diff
@@ -0,0 +1,238 @@
+    table::{Decompress, Table},
+    transaction::{DbTx, DbTxMut},
+    BlockNumberList, DatabaseError,
+};
+use reth_etl::Collector;
+use reth_primitives::BlockNumber;
+use std::{collections::HashMap, hash::Hash, ops::RangeBounds};
+use tracing::info;
+
+/// Number of entries before pushing indices from cache to [`Collector`]
+const DEFAULT_CACHE_THRESHOLD: usize = 200_000;
+
+/// Collects all history (`H`) indices for a range of changesets (`CS`) and stores them in a
+/// [`Collector`].
+///
+/// ## Process
+/// The function utilizes a `HashMap` cache with a structure of `PartialKey` (`P`) (Address or
+/// Address.StorageKey) to `BlockNumberList`. When the cache exceeds its capacity, its contents are
+/// moved to a [`Collector`]. Here, each entry's key is a concatenation of `PartialKey` and the
+/// highest block number in its list.
+///
+/// ## Example
+/// 1. Initial Cache State: `{ Address1: [1,2,3], ... }`
+/// 2. Cache is flushed to the `Collector`.
+/// 3. Updated Cache State: `{ Address1: [100,300], ... }`
+/// 4. Cache is flushed again.
+///
+/// As a result, the `Collector` will contain entries such as `(Address1.3, [1,2,3])` and
+/// `(Address1.300, [100,300])`. The entries may be stored across one or more files.
+pub(crate) fn collect_history_indices<TX, CS, H, P>(
+    tx: &TX,
+    range: impl RangeBounds<CS::Key>,
+    sharded_key_factory: impl Fn(P, BlockNumber) -> H::Key,
+    partial_key_factory: impl Fn((CS::Key, CS::Value)) -> (u64, P),
+) -> Result<Collector<H::Key, H::Value>, StageError>
+where
+    TX: DbTxMut + DbTx,
+    CS: Table,
+    H: Table<Value = BlockNumberList>,
+    P: Eq + Hash,
+{
+    let mut changeset_cursor = tx.cursor_read::<CS>()?;
+
+    let mut collector = Collector::new(500 * 1024 * 1024, None);
+    let mut cache: HashMap<P, Vec<u64>> = HashMap::new();
+    let mut entries = 0;
+
+    let mut collect = |cache: HashMap<P, Vec<u64>>| {
+        for (key, indice_list) in cache {
+            let last = indice_list.last().expect("qed");
+            collector.insert(
+                sharded_key_factory(key, *last),
+                BlockNumberList::new_pre_sorted(indice_list),
+            )?;
+        }
+        Ok::<(), StageError>(())
+    };
+
+    // observability
+    let total_changesets = tx.entries::<CS>()?;
+    let interval = (total_changesets / 100).max(1);
+
+    for entry in changeset_cursor.walk_range(range)? {
+        let (index, key) = partial_key_factory(entry?);
+        cache.entry(key).or_default().push(index);
+
+        if entries > 0 && entries % interval == 0 && total_changesets > 100 {
+            info!(target: "sync::stages::index_history", progress = %format!("{:.2}%", (entries as f64 / total_changesets as f64) * 100.0), "Collecting indices");
+        }
+
+        entries += 1;
+        if entries > DEFAULT_CACHE_THRESHOLD {
+            entries = 0;
+
+            collect(cache)?;
+            cache = HashMap::default();
```

Comment:
should we just clean the hashmap instead of deallocating the capacity, given that we already allocated `DEFAULT_CACHE_THRESHOLD` elements?